### PR TITLE
HPCC-23004 Update Std.DataPatterns to version 1.5.7

### PR DESCRIPTION
**Pull request by dcamper**
Signed-off-by: Dan S. Camper <dan.camper@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:

Manual testing; unit testing using tests from bundle version

### CODE REVIEW

**Comment by dcamper** [comment]
@GordonSmith Please review

File: ecllibrary/std/DataPatterns/Profile.ecl [hunk 1]
```diff
@@ -159,12 +159,12 @@
  *                              quartiles               numeric_lower_quartile
  *                                                      numeric_median
  *                                                      numeric_upper_quartile
- *                              correlations            numeric_correlations
+ *                              %correlations%            numeric_correlations
```
Comment (GordonSmith) [inline ecllibrary/std/DataPatterns/Profile.ecl]: Was this changed on purpose (looks like a global change)?

File: ecllibrary/std/DataPatterns/Profile.ecl [hunk 2]
```diff
@@ -159,12 +159,12 @@
...
  *                          To omit the output associated with a single keyword,
  *                          set this argument to a comma-delimited string
  *                          containing all other keywords; note that the
  *                          is_numeric output will appear only if min_max,
- *                          mean, std_dev, quartiles, or correlations features
+ *                          mean, std_dev, quartiles, or %correlations% features
```
Comment (GordonSmith) [inline ecllibrary/std/DataPatterns/Profile.ecl]: Was this changed on purpose (looks like a global change)?

File: ecllibrary/std/DataPatterns/Profile.ecl (hunk 1 above)
Comment (dcamper) [reply ecllibrary/std/DataPatterns/Profile.ecl]: There was a correction of an incorrect global find-and-replace that resulted in the %correlations% string appearing.  I missed that one during the correction; will fix.

**Review by GordonSmith — COMMENTED** [review]
@dcamper I am not sure I am qualified to review the actual ECL - but am I correct in saying its _only_ a change to the internal attribute names to avoid collisions? 

Are there any additional features included (which would indicate it might need to wait util 7.8)?

Also you would typically issue the PR to candidate-7.6.x and it will get included in the next point release.

Or target candidate-7.8.x / master (if it has features).

**Comment by dcamper** [comment]
@GordonSmith Changes pushed and squashed.

Only the atttribute renaming and general bug fixes are included; no additional features.

I targeted 7.6.4 because the candidate already existed and I'm hoping to get it in that release due to the recent exposure at the conference.

**Comment by dcamper** [comment]
@GordonSmith I failed to request that you explicitly test the ECL Watch integration.  Just want to make sure all that still works.

**Comment by richardkchapman** [comment]
@dcamper Please target 7.6.x rather than any specific 7.6 series version

**Comment by dcamper** [comment]
Rebased to 7.6.x.

File: ecllibrary/std/DataPatterns/Profile.ecl [hunk 3]
```diff
@@ -190,14 +190,12 @@ EXPORT Profile(inFile,
...
                sampleSize = 100,
                lcbLimit = 64) := FUNCTIONMACRO
-    IMPORT STD;
     LOADXML('<xml/>');
 
     #UNIQUENAME(temp);                      // Ubiquitous "contains random things"
     #UNIQUENAME(scalarFields);              // Contains a delimited list of scalar attributes (full names) along with their indexed positions
     #UNIQUENAME(explicitScalarFields);      // Contains a delimited list of scalar attributes (full names) without indexed positions
     #UNIQUENAME(childDSFields);             // Contains a delimited list of child dataset attributes (full names) along with their indexed positions
-    #UNIQUENAME(explicitAllFields);         // Contains a delimited list of all attributes (full names) without indexed positions
```
Comment (dcamper) [inline ecllibrary/std/DataPatterns/Profile.ecl]: Diff: Removed unused attribute

File: ecllibrary/std/DataPatterns/Profile.ecl [hunk 4]
```diff
@@ -302,6 +326,19 @@ EXPORT Profile(inFile,
                     #APPEND(childDSFields, ',')
                 #END
                 #APPEND(childDSFields, %'fieldCount'% + ':' + %'fullName'%)
+                // Extract the child dataset into its own attribute so we can more easily
```
Comment (dcamper) [inline ecllibrary/std/DataPatterns/Profile.ecl]: Diff: New code as workaround, allowing a child dataset within a child record to be profiled.

File: ecllibrary/std/DataPatterns/Profile.ecl [hunk 5]
```diff
@@ -320,107 +357,103 @@ EXPORT Profile(inFile,
 
     // Collect the gathered full attribute names so we can walk them later
     #SET(explicitScalarFields, REGEXREPLACE('\\d+:', %'scalarFields'%, ''));
-    #SET(explicitAllFields, %'scalarFields'%);
```
Comment (dcamper) [inline ecllibrary/std/DataPatterns/Profile.ecl]: Diff: This attribute was removed as unused elsewhere.

File: ecllibrary/std/DataPatterns/Profile.ecl [hunk 6]
```diff
@@ -598,90 +647,103 @@ EXPORT Profile(inFile,
...
+                                                                            #ELSE
+                                                                                FALSE
+                                                                            #END
+                                                },
+                                                _inFile.#EXPAND(%'namePrefix'% + %'@name'%),
+                                                LOCAL
+                                            ),
+                                            TRANSFORM(%DataInfoRec%, SELF := LEFT)
+                                    ),
+                                DATASET
```
Comment (dcamper) [inline ecllibrary/std/DataPatterns/Profile.ecl]: Diff: This is the else part of an IF() above, handling the case where a child dataset contains no records anywhere in the outer dataset.

File: ecllibrary/std/DataPatterns/Profile.ecl [hunk 7]
```diff
@@ -719,109 +784,109 @@ EXPORT Profile(inFile,
...
+                                (LEFT.type_flag & %DataTypeEnum%.FloatingPoint) != 0                                    =>  'real' + IF(LEFT.data_length < 8, '4', '8'),
+                                (LEFT.type_flag & %DataTypeEnum%.ExpNotation) != 0                                      =>  'real8',
                                 REGEXFIND('utf', LEFT.given_attribute_type)                                             =>  LEFT.given_attribute_type,
                                 REGEXREPLACE('\\d+$', TRIM(LEFT.given_attribute_type), '') + IF(LEFT.data_length > 0 AND (LEFT.data_length < (LEFT.min_data_length * 1000)), (STRING)LEFT.data_length, '')
                             ),
                         SELF := LEFT
                     )
             );
 
-        // Record definition for mode values that we'll be returning
```
Comment (dcamper) [inline ecllibrary/std/DataPatterns/Profile.ecl]: Diff: This definition was moved to earlier in the code

File: ecllibrary/std/DataPatterns/Profile.ecl [hunk 8]
```diff
@@ -1315,18 +1397,19 @@ EXPORT Profile(inFile,
                                 SELF := RIGHT,
                                 SELF := LEFT
                             ),
-                        KEEP(1), SMART
+                        LEFT OUTER, KEEP(1), SMART
```
Comment (dcamper) [inline ecllibrary/std/DataPatterns/Profile.ecl]: Diff: Add missing JOIN flag

File: ecllibrary/std/DataPatterns/Profile.ecl [hunk 9]
```diff
@@ -1442,14 +1529,18 @@ EXPORT Profile(inFile,
             #IF(%'dsNameValue'% != '')
                 #SET(numValue, REGEXFIND('^(\\d+):', %'dsNameValue'%, 1))
                 #SET(nameValue, REGEXFIND(':([^:]+)$', %'dsNameValue'%, 1))
-                + _Profile
+                // The child dataset should have been extracted into its own
+                // local attribute; reference it during our call to the inner
+                // profile function macro
+                #SET(temp, #MANGLE(%'nameValue'%))
```
Comment (dcamper) [inline ecllibrary/std/DataPatterns/Profile.ecl]: Diff: Use previously-created local attribute, referencing a child dataset, to work around ECL problems with dot-syntax referencing in some circumstances

**Comment by dcamper** [comment]
@richardkchapman 98% of the (many) changes within this PR involve using #UNIQUENAME to avoid name collision with calling code and to allow multiple Profile() calls within a single ECL job.  I have went through the diff and added notes to the few changes that are something else, in an effort to aid review.  Who should review this PR?

**Comment by ghalliday** [comment]
@ghalliday I will take a look at it, but not today...

**Review by ghalliday — COMMENTED** [review]
Changes all look fine.  Thanks for the comments

**Comment by dcamper** [comment]
Thank you for the review, Gavin.

@richardkchapman Please merge.

**Pull request merged**